### ISSUE TITLE
delete-dups call in projectile-ag causes "Attempt to modify read-only object" error

### ISSUE THREAD

**Issue by nwidger** [bbatsov/projectile#1162]
I get the following error when I try to run `projectile-ag`: 

```
Debugger entered--Lisp error: (error "Attempt to modify read-only object" ("SCCS" "RCS" "CVS" "MCVS" ".src" ".svn" ".git" ".hg" ".bzr" "_MTN" "_darcs" "{arch}"))
  delete-dups(("*.svn-base" "*.o" "*.pyc" ".gitignore" "TAGS" ".svn/" ".idea/" ".eunit/" ".git/" ".hg/" ".fslckout/" ".bzr/" "_darcs/" ".tox/" "Godeps/" ".#*" "*~" "*.bin" "*.lbin" "*.so" "*.a" "*.ln" "*.blg" "*.bbl" "*.elc" "*.lof" "*.glo" "*.idx" "*.lot" "*.fmt" "*.tfm" "*.class" "*.fas" "*.lib" "*.mem" "*.x86f" "*.sparcf" "*.dfsl" "*.pfsl" "*.d64fsl" "*.p64fsl" "*.lx64fsl" "*.lx32fsl" "*.dx64fsl" "*.dx32fsl" "*.fx64fsl" "*.fx32fsl" "*.sx64fsl" "*.sx32fsl" "*.wx64fsl" ...))
  projectile-ag("" nil)
  funcall-interactively(projectile-ag "" nil)
  call-interactively(projectile-ag nil nil)
  command-execute(projectile-ag)
```

If I remove the `delete-dups` call from `projectile-ag`, the error goes away.  I believe commit https://github.com/bbatsov/projectile/commit/916c87f2ffce267b770a3121aa261a4d083a401c is to be the culprit.

I am running Emacs 25.2.1 on CentOS 7, projectile 20170727.2351 and ag 20170712.1549 (both from MELPA).  Please let me know if I can provide any more information.

**Comment by czw** [comment]
Replicated with Projectile 20170827.2053 as well. Environment: Emacs 25.2.1, macOS 10.3.3.

**Comment by danielbarbarito** [comment]
Replicated with Projectile 0.15.0-cvs. This only happens when I search in non-git projects. As soon as I add a `.git` folder (even if it is completely empty) everything works.

**Comment by nwidger** [comment]
@danielbarbarito Is removing the call to `delete-dups` the right solution?  Presumably having duplicates in `ag-ignore-list` was a causing issues for someone, or else the call wouldn't have been added by https://github.com/bbatsov/projectile/commit/916c87f2ffce267b770a3121aa261a4d083a401c.  After digging around a bit with `apropos-value`, I believe the `read-only object` that projectile tries to modify (at least using the stacktrace I included above) is `grep-find-ignored-directories`.  It makes sense that adding a `.git` folder would paper over the issue, since `projectile-ag` only references `grep-find-ignored-directories` if `projectile-project-vcs` returns `git`.

**Comment by nwidger** [comment]
As an extra data point, from my own testing simply removing the reference to `grep-find-ignored-directories` in `projectile-ag` seems to resolve the issue for non-Git projects.  That isn't a proper solution, but it does narrow the issue down to the handling of `grep-find-ignored-directories` somehow.

**Comment by danielbarbarito** [comment]
@nwidger Don't get me wrong, removing the delete-dups call is definitely not the right solution but it solved the issue for me and that's all I needed. Good work on tracing the issue down to grep-find-ignored-directories though!

**Comment by bbatsov** [comment]
I'm really puzzled about this - I don't see anything that should returning a read-only list in the code.

**Comment by nwidger** [comment]
@bbatsov I completely agree, nothing in the `grep` package where `grep-find-ignored-directories` is declared seems a likely culprit.  Perhaps some other package is mucking with it somehow?  I also can't find anything that explains how to make a variable in Emacs Lisp read-only.  The only thing my searching has turned up is lots of people saying not to use `delete-dups` because it's destructive and you should know what you're doing when you use it.

In general, it would be great to get to the bottom of this and determine the underlying cause, but I really just want to be able to use `projectile-ag` again.  Up until this bug started happening I was using it constantly and find it extremely useful.

**Comment by dedi** [comment]
Hi. I think I might have a workaround for this issue, or at least a clue. Investigating a little, looking at Emacs version 25.3, I noticed that, in vc-hooks.el, vc-directory-exclusion-list is declared using purecopy, so it is marked as read-only. Then, in project.el function project-ignores, grep-find-ignored-files is set using nconc. nconc doesn't alter the last parameter, so it is appended to the list as is, meaning the end of the list is at read-only memory. Then, in projectile-ag function, the ag-ignore-list is created using append, with grep-find-ignored-directories as the last parameter. Again, append copies the last parameter as is, so the end of the list is still read-only.
A workaround is to make sure grep-find-ignored-directories is copied. I did it by adding an empty list as the last parameter to append, like so:
```
...
                                    (unless (eq (projectile-project-vcs) 'git)
                                      (append (projectile-ignored-files-rel)
                                              (projectile-ignored-directories-rel)
                                              grep-find-ignored-files
                                              grep-find-ignored-directories '()))))))
...
```
 This seemed to solve the problem for me. I'm guessing moving grep-find-ignored-directories to before grep-find-ignored-files would work too, but I thought my way is better, because it doesn't assume any of the variables is writable. There are probably other ways to force the copy of the value of ,grep-find-ignored-directories but I am not an ELisp person.
